### ISSUE TITLE
State copy or set messes up results

### ISSUE THREAD

**Issue by niansa** [ggerganov/llama.cpp#4005]
# Prerequisites

Please answer the following questions for yourself before submitting an issue.

- [x] I am running the latest code. Development is very rapid so there are no tagged versions as of now.
- [x] I carefully followed the [README.md](https://github.com/ggerganov/llama.cpp/blob/master/README.md).
- [x] I [searched using keywords relevant to my issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/filtering-and-searching-issues-and-pull-requests) to make sure that I am creating a new issue that is not already open (or closed).
- [x] I reviewed the [Discussions](https://github.com/ggerganov/llama.cpp/discussions), and have a new bug or useful enhancement to share.

# Expected Behavior

Using `llama_copy_state_data`, then evaluating tokens, then using `llama_set_state_data` should return the model back to its previous state.

# Current Behavior

Generation goes off badly.

# Environment and Context

* Physical (or virtual) hardware you are using, e.g. for Linux:

```
Architecture:            x86_64
  CPU op-mode(s):        32-bit, 64-bit
  Address sizes:         44 bits physical, 48 bits virtual
  Byte Order:            Little Endian
CPU(s):                  8
  On-line CPU(s) list:   0-7
Vendor ID:               AuthenticAMD
  Model name:            AMD Custom APU 0405
    CPU family:          23
    Model:               144
    Thread(s) per core:  2
    Core(s) per socket:  4
    Socket(s):           1
    Stepping:            2
    Frequency boost:     enabled
    CPU(s) scaling MHz:  55%
    CPU max MHz:         2800,0000
    CPU min MHz:         1700,0000
    BogoMIPS:            5599,65
    Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 h
                         t syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_ap
                         icid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx 
                         f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs 
                         skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate
                          ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 cqm rdt_a rdseed adx smap clflushopt
                          clwb sha_ni xsaveopt xsavec xgetbv1 cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irper
                         f xsaveerptr rdpru wbnoinvd cppc arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid 
                         decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip rdpid overflow_rec
                         ov succor smca sev sev_es
Virtualization features: 
  Virtualization:        AMD-V
Caches (sum of all):     
  L1d:                   128 KiB (4 instances)
  L1i:                   128 KiB (4 instances)
  L2:                    2 MiB (4 instances)
  L3:                    4 MiB (1 instance)
NUMA:                    
  NUMA node(s):          1
  NUMA node0 CPU(s):     0-7
Vulnerabilities:         
  Gather data sampling:  Not affected
  Itlb multihit:         Not affected
  L1tf:                  Not affected
  Mds:                   Not affected
  Meltdown:              Not affected
  Mmio stale data:       Not affected
  Retbleed:              Mitigation; untrained return thunk; SMT enabled with STIBP protection
  Spec rstack overflow:  Mitigation; safe RET
  Spec store bypass:     Mitigation; Speculative Store Bypass disabled via prctl
  Spectre v1:            Mitigation; usercopy/swapgs barriers and __user pointer sanitization
  Spectre v2:            Mitigation; Retpolines, IBPB conditional, STIBP always-on, RSB filling, PBRSB-eIBRS Not affected
  Srbds:                 Not affected
  Tsx async abort:       Not affected
```

* Operating System, e.g. for Linux:

```
Linux nils-jupiter 6.5.0-2-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.5.6-1 (2023-10-07) x86_64 GNU/Linux
```

* SDK version, e.g. for Linux:

```
g++ (Debian 13.2.0-5) 13.2.0
Copyright (C) 2023 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

# Failure Information (for bugs)

See below

# Steps to Reproduce

Pseudo code:

    LM::Inference::State sv;
    model->append("This is a");
    model->copy_state(sv);
    model->set_state(sv);
    puts("A begin");
    model->evaluate(" chair that");
    model->generate(show_result);
    puts("A end");
    model->set_state(sv);
    puts("B begin");
    model->evaluate(" car that");
    model->generate(show_result);
    puts("B end");

# Failure Logs

Pseudo code output:

```
A begin
 can be used in various settings, It is designed to be used in different settings.A end
B begin
 way that it is made from a variety of settings.
B end
```

**Comment by WeirdConstructor** [comment]
@niansa Are you running with offloaded GPU layers? Is it maybe the same issue as described in #2422?

**Comment by niansa** [comment]
Nope, just default CPU

**Comment by blackhole89** [comment]
I think I can reproduce the problem even with a state that was saved with `llama_copy_state_data` from a fresh context with nothing decoded yet: `llama_set_state_data` with such a state ought to "factory-reset" the context, but it doesn't actually and instead logits that come out of the resulting state are bogus, and moreover change with repeated iterations of set_state_data and decoding identical batches.

The problem seems to boil down to something about how the KV cache is restored or not restored: adding and invoking a helper function 
```cpp
void llama_reinit_kv_cache(struct llama_context *ctx, struct llama_context_params   params)
{
    ctx->kv_self = llama_kv_cache();
    llama_kv_cache_init(ctx->kv_self, ctx->model, params.type_k, params.type_v,
                ctx->cparams.n_ctx, ctx->cparams.offload_kqv);
}
```
before set_state_data is sufficient to make subsequent outputs consistent with repeated iterations of set_state_data. I will investigate further.

Relatedly, but insufficiently to fix, in `llama_set_state_data`'s `//set logits` block, the logic of 
```
if (logits_size) {
        ctx->logits.resize(logits_size);
```
seems wrong: it is possible (as in my example) that a state with logits_size 0 was stored, and then ctx->logits should in fact be emptied.

**Comment by ggerganov** [comment]
Without looking much in details yet, #4820 might provide some additional context

p.s. @blackhole89 good to see you again :)

**Comment by blackhole89** [comment]
Thanks for the pointer, I'll check out the discussion there! Also thanks for the welcome, it's great to see how much the project has grown :)

**Comment by dfriehs** [comment]
> [...] seems wrong: it is possible (as in my example) that a state with logits_size 0 was stored, and then ctx->logits should in fact be emptied.

While I agree that `ctx->logits` should be emptied when resetting state to a blank slate I'm not sure if there are any benefits to doing so:
- Logits are never read during calls to `llama_decode` and always resized to non-zero sizes
- Logits frequently already contain garbage when using the new `llama_batch` API as only positions with `batch.logits[idx] == true` are updated
- Debug guards (`ctx.logits_valid`) are not serialized at all and will currently mistakingly assert in debug builds even on now "valid" positions after deserialization

This makes **all** sampling after deserialization only safe if you are **sure** that the logits at your chosen index have been calculated before saving (and never in debug builds). Calls to `llama_decode` should be completely unaffected.

**Comment by niansa** [comment]
So the bug is not just something weird on my end. Phew.

It is currently breaking a lot of my projects, really looking forward to a fix. If there is something I can do to help debug this issue I'd be happy to help.
